In [2]:
from pathlib import Path
answers_dir = Path(r"../CERT_DATA/ftp.sei.cmu.edu/pub/cert-data/answers")
dataset_dir = Path(r"../CERT_DATA/ftp.sei.cmu.edu/pub/cert-data/r5.2")

main_answers_file = answers_dir / "insiders.csv"

# Detail of malicious user

In [4]:
insider_files = list(answers_dir.glob('**/r5.2*.csv'))
# insider_files

In [8]:
df_ls = []
import pandas as pd
for filename in insider_files:
    df = pd.read_csv(filename, names=list(range(13)))
    df_ls.append(df)

df = pd.concat(df_ls, axis=0, ignore_index=True)
df

0                         1                    2        3        4   \
0      logon  {K3V4-Y4OK65SI-1583GEOQ}  10/23/2010 01:34:19  AAM0658  PC-9923   
1     device  {H1L0-X7RH83FI-5967VUQY}  10/23/2010 06:18:48  AAM0658  PC-9923   
2       http  {Y4Q9-U5VQ11UG-1279ZPTL}  10/23/2010 06:26:01  AAM0658  PC-9923   
3     device  {W7C0-G1SW41KB-1991BUTL}  10/23/2010 06:26:48  AAM0658  PC-9923   
4      logon  {C7T3-B4IY16JF-6945LOFC}  10/23/2010 06:28:17  AAM0658  PC-9923   
5      logon  {S8Q1-Z1OG01ED-8756DWPR}  10/29/2010 00:06:47  AAM0658  PC-9923   
6     device  {S1J0-Z2ZX87IX-2607ZUXY}  10/29/2010 01:30:10  AAM0658  PC-9923   
7       http  {Z8E0-S5AG27DJ-3234IKCE}  10/29/2010 01:36:11  AAM0658  PC-9923   
8     device  {V6D8-B3FZ26HT-3583PJAI}  10/29/2010 05:00:19  AAM0658  PC-9923   
9      logon  {J6L0-W5OJ26KB-8630RYVJ}  10/29/2010 05:23:28  AAM0658  PC-9923   
10     logon  {H5M9-E4GU47TS-2991DMNK}  09/21/2010 01:16:22  BLS0678  PC-6031   
11    device  {S1T0-E5SZ35TF-5143KLFY}  09/21/2010 02:46:59  BLS0678  PC-6031   
12      http  {N7J5-Z3BW40PP-9252FCEK}  09/21/2010 02:47:56  BLS0678  PC-6031   
13    device  {R4T0-A0VP51YA-2744THEC}  09/21/2010 02:48:37  BLS0678  PC-6031   
14     logon  {M3M4-C8YB70JN-4776PTQV}  09/21/2010 03:44:49  BLS0678  PC-6031   
15     logon  {Y3V9-J3OX51PK-1617PXNW}  09/29/2010 17:36:12  BLS0678  PC-6031   
16    device  {F0D4-X4BP77OM-1802ULCG}  09/29/2010 18:52:55  BLS0678  PC-6031   
17      http  {X8W9-N2TW91WV-9895JKUB}  09/29/2010 19:01:18  BLS0678  PC-6031   
18    device  {E3P0-I0CM71JW-7152TNLM}  09/29/2010 19:05:29  BLS0678  PC-6031   
19     logon  {A4M8-E2JY79ZI-1744AMBM}  09/30/2010 04:48:19  BLS0678  PC-6031   
20     logon  {K1B2-N6OZ19XO-7148MLSP}  07/13/2010 20:15:23  BIH0745  PC-2661   
21    device  {Y3M2-Z6XM78GC-6283GEZU}  07/13/2010 21:06:06  BIH0745  PC-2661   
22      http  {O4T8-D4QZ55FR-8512IAJW}  07/13/2010 21:06:44  BIH0745  PC-2661   
23    device  {S7Z7-F8GG07WE-5470ZMDD}  07/13/2010 21:09:08  BIH0745  PC-2661   
24     logon  {I4V0-J7OH43HZ-7956NEXE}  07/13/2010 21:20:44  BIH0745  PC-2661   
25     logon  {L1A0-X3OD03QV-9450KGJI}  11/25/2010 06:35:11  FTM0406  PC-1786   
26    device  {Y5K5-C6AF28XA-5238BVTN}  11/25/2010 06:45:46  FTM0406  PC-1786   
27      http  {N5G9-W9SS50SL-9399ZFUG}  11/25/2010 06:45:57  FTM0406  PC-1786   
28    device  {S0F7-W2LZ64WJ-4108ZIKN}  11/25/2010 06:46:03  FTM0406  PC-1786   
29     logon  {M4Y1-P1MF23RY-1605UHAS}  11/25/2010 06:51:18  FTM0406  PC-1786   
...      ...                       ...                  ...      ...      ...   
7293  device  {F0P9-A1KZ99RF-7804VUPC}  10/20/2010 16:07:43  HXL0968  PC-0623   
7294  device  {R0K0-W9PG22WU-1870RSXQ}  10/20/2010 16:13:49  HXL0968  PC-0623   
7295  device  {E9P6-E0NH58EG-1521THQM}  10/20/2010 16:16:39  HXL0968  PC-0623   
7296  device  {G6W2-L4KI05IK-5038CJVG}  10/20/2010 16:17:21  HXL0968  PC-0623   
7297  device  {X4R2-M1YR64WR-4796PKRL}  10/25/2010 10:12:43  HXL0968  PC-0623   
7298  device  {K8L3-D2PP50WX-6785HIVT}  10/25/2010 10:15:02  HXL0968  PC-0623   
7299  device  {O0F8-S6NR58QE-5391JBFV}  10/25/2010 10:19:16  HXL0968  PC-0623   
7300  device  {Z7F1-Y7RF75AP-5043RLTM}  10/25/2010 10:19:35  HXL0968  PC-0623   
7301  device  {E6W7-X9ST43JK-5328WBSK}  10/25/2010 13:41:18  HXL0968  PC-0623   
7302  device  {L0N9-K9AH69CU-8738ZKJW}  10/25/2010 13:42:39  HXL0968  PC-0623   
7303  device  {C8W6-B3MN68SN-6557QMDD}  10/25/2010 13:43:19  HXL0968  PC-0623   
7304  device  {D5X4-I3LJ23BP-6879CTEJ}  10/25/2010 13:44:09  HXL0968  PC-0623   
7305  device  {Z3E4-T8XT30EY-5355UYBM}  10/25/2010 13:44:34  HXL0968  PC-0623   
7306  device  {Y9F9-L8BV40TH-3284OKIN}  10/25/2010 13:54:51  HXL0968  PC-0623   
7307  device  {I2V2-B5PI66GC-4209VEIH}  10/25/2010 14:35:50  HXL0968  PC-0623   
7308  device  {C9N1-I0GX23LA-6310TRZI}  10/25/2010 14:38:25  HXL0968  PC-0623   
7309  device  {U1L9-G0SI43HV-1554PPAJ}  10/28/2010 07:27:17  HXL0968  PC-0623   
7310  device  {O1L9-A4IZ62UC-4403REJU}  1

In [10]:
main_df = pd.read_csv(main_answers_file)
main_df = main_df[main_df['dataset'] == 5.2]
main_df

total of  99  malicious users


array(['ALT1465', 'ALW0764', 'AYG1697', 'DAS1320', 'DNJ0740', 'ELT1370',
       'EPG1196', 'ETW0002', 'FZG0389', 'GFM1815', 'IHC0561', 'ISW0738',
       'JKB0287', 'JUP1472', 'KBC1390', 'KEW0198', 'MIB0203', 'NIV1608',
       'PBC0077', 'PTH0005', 'REF1924', 'SAF1942', 'SLL0193', 'TMC0934',
       'VAH1292', 'WHB1247', 'WSK1857', 'WWW0701', 'ZKP0542', 'VCF1602',
       'KSS1005', 'MCP0611', 'TNB1616', 'ZIE0741', 'HXP0976', 'HIS1394',
       'JAL0811', 'ITA0159', 'CIF1430', 'RRS0056', 'MDS0680', 'TMT0851',
       'OSS1463', 'HSN0675', 'WDT1634', 'OKM1092', 'CKP0630', 'ICB1354',
       'TRC1838', 'NAH1366', 'BYO1846', 'GWG0497', 'MGB1235', 'CHP1711',
       'SNK1280', 'LVF1626', 'HMS1658', 'DCC1119', 'SIS0042', 'ACA1126',
       'CRD0272', 'CWW1120', 'ELM1123', 'GKW0043', 'KCM0466', 'LAH0463',
       'MPF0690', 'VRP0267', 'ZEH0685', 'ACM1770', 'AMC1883', 'ASH0298',
       'ASH0458', 'BDP0096', 'CBW1826', 'CGF1056', 'CPD1525', 'DIM0330',
       'DKG0161', 'DQM0809', 'GCD0194', 'GMB0400', 

In [ ]:
malicious_users = main_df.user.unique()
print("total of ", len(malicious_users)," malicious users")
# malicious_users

In [23]:
print("total of ", len(main_df.scenario.unique())," malicious senerios")
print("senerios 1: ", len(main_df[main_df['scenario'] == 1]))
print("senerios 2: ", len(main_df[main_df['scenario'] == 2]))
print("senerios 3: ", len(main_df[main_df['scenario'] == 3]))
print("senerios 4: ", len(main_df[main_df['scenario'] == 4]))

total of  4  malicious senerios
senerios 1:  29
senerios 2:  30
senerios 3:  10
senerios 4:  30


In [25]:
# main_df[main_df['scenario'] == 1]

dataset  scenario             details     user                start  \
82       5.2         1  r5.2-1-ALT1465.csv  ALT1465  08/13/2010 22:16:10   
83       5.2         1  r5.2-1-ALW0764.csv  ALW0764  02/03/2011 07:09:09   
84       5.2         1  r5.2-1-AYG1697.csv  AYG1697  10/25/2010 23:19:22   
85       5.2         1  r5.2-1-DAS1320.csv  DAS1320  07/09/2010 01:14:15   
86       5.2         1  r5.2-1-DNJ0740.csv  DNJ0740  10/04/2010 19:56:42   
87       5.2         1  r5.2-1-ELT1370.csv  ELT1370  02/03/2011 18:23:53   
88       5.2         1  r5.2-1-EPG1196.csv  EPG1196  07/15/2010 03:00:42   
89       5.2         1  r5.2-1-ETW0002.csv  ETW0002  03/17/2011 00:59:27   
90       5.2         1  r5.2-1-FZG0389.csv  FZG0389  12/10/2010 21:50:34   
91       5.2         1  r5.2-1-GFM1815.csv  GFM1815  07/16/2010 22:00:57   
92       5.2         1  r5.2-1-IHC0561.csv  IHC0561  09/02/2010 06:10:00   
93       5.2         1  r5.2-1-ISW0738.csv  ISW0738  10/27/2010 21:38:38   
94       5.2         1  r5.2-1-JKB0287.csv  JKB0287  09/09/2010 02:50:22   
95       5.2         1  r5.2-1-JUP1472.csv  JUP1472  02/11/2011 22:17:37   
96       5.2         1  r5.2-1-KBC1390.csv  KBC1390  07/16/2010 02:18:07   
97       5.2         1  r5.2-1-KEW0198.csv  KEW0198  07/07/2010 18:50:50   
98       5.2         1  r5.2-1-MIB0203.csv  MIB0203  10/13/2010 23:51:44   
99       5.2         1  r5.2-1-NIV1608.csv  NIV1608  02/05/2011 01:22:24   
100      5.2         1  r5.2-1-PBC0077.csv  PBC0077  07/22/2010 02:04:28   
101      5.2         1  r5.2-1-PTH0005.csv  PTH0005  01/31/2011 22:25:29   
102      5.2         1  r5.2-1-REF1924.csv  REF1924  10/14/2010 19:08:01   
103      5.2         1  r5.2-1-SAF1942.csv  SAF1942  07/21/2010 00:37:49   
104      5.2         1  r5.2-1-SLL0193.csv  SLL0193  08/27/2010 18:57:31   
105      5.2         1  r5.2-1-TMC0934.csv  TMC0934  02/17/2011 06:53:38   
106      5.2         1  r5.2-1-VAH1292.csv  VAH1292  04/29/2011 22:01:07   
107      5.2         1  r5.2-1-WHB1247.csv  WHB1247  02/14/2011 21:50:42   
108      5.2         1  r5.2-1-WSK1857.csv  WSK1857  03/04/2011 00:09:34   
109      5.2         1  r5.2-1-WWW0701.csv  WWW0701  03/23/2011 06:54:24   
110      5.2         1  r5.2-1-ZKP0542.csv  ZKP0542  01/12/2011 20:00:15   

                     end  
82   08/20/2010 04:01:32  
83   02/05/2011 06:12:51  
84   10/26/2010 00:06:28  
85   07/15/2010 01:38:58  
86   10/05/2010 01:01:37  
87   02/10/2011 02:58:28  
88   07/20/2010 00:19:48  
89   03/17/2011 06:52:18  
90   12/17/2010 06:53:00  
91   07/16/2010 23:01:54  
92   09/10/2010 08:28:40  
93   11/03/2010 07:01:27  
94   09/15/2010 02:29:40  
95   02/19/2011 04:15:38  
96   07/22/2010 00:50:53  
97   07/14/2010 06:27:14  
98   10/19/2010 02:45:24  
99   02/12/2011 06:15:24  
100  07/31/2010 02:28:03  
101  02/04/2011 05:42:33  
102  10/19/2010 22:24:04  
103  07/23/2010 02:12:05  
104  09/03/2010 00:24:08  
105  02/22/2011 21:01:13  
106  05/07/2011 04:04:19  
107  02/19/2011 01:07:48  
108  03/11/2011 06:52:02  
109  03/30/2011 05:40:52  
110  01/15/2011 04:17:00

# Detail of Normal User

In [39]:
logon_df = pd.read_csv(dataset_dir / 'logon.csv')
# logon_df

In [29]:
print("total of user:", len(logon_df['user'].unique()))
print("total of Normal User: ", len(logon_df['user'].unique()) - len(malicious_users))

total of user: 2000
total of Normal User:  1901


# Summary until now
|  total user | senerio 1 | senerio 2 | senerio 3 | senerio 4| Normal user |
| --- | --- | --- | -- | -- | -- |
| 2000 | 29 | 30 | 10 | 30 | 1901 |

# Detail of log quantities

In [41]:
logon_df['date'] = pd.to_datetime(logon_df.date, format='%m/%d/%Y %H:%M:%S')

logon_df['day'] =logon_df['date'].dt.floor('D')
logon_df

id                date     user       pc  \
0        {Q4D5-W4HH44UC-5188LWZK} 2010-01-02 02:24:51  JBI1134  PC-0168   
1        {G7V0-S4TP95SA-9203AOGR} 2010-01-02 02:38:28  JBI1134  PC-0168   
2        {B4U7-K4DB84LM-2657VBFY} 2010-01-02 04:55:52  JBI1134  PC-2320   
3        {K2E1-W7VG04OA-3686THSV} 2010-01-02 05:02:28  JBI1134  PC-2320   
4        {Z7B2-Z8AS75YE-4485NAFE} 2010-01-02 06:35:00  HMI1448  PC-9352   
5        {H2N5-X1QK85TW-4800RFFG} 2010-01-02 06:36:55  JBI1134  PC-5439   
6        {S4B7-R6UW34WW-9244WTVD} 2010-01-02 06:38:42  JBI1134  PC-5439   
7        {K5P1-Y6HZ87QS-5849WSYS} 2010-01-02 06:46:00  KMC1934  PC-2683   
8        {P6B3-G4RP59TS-5233ZZFI} 2010-01-02 06:50:00  JBI1134  PC-7235   
9        {R7Y5-U1VO87KN-5237KNZT} 2010-01-02 06:55:00  HRE1950  PC-8025   
10       {T1U7-I7JB26LP-7140HNMC} 2010-01-02 06:56:00  HBP0009  PC-8466   
11       {U2A0-J3ZO81IH-6510VANL} 2010-01-02 06:58:00  OCW1127  PC-7239   
12       {A3Q7-G9CV17YG-8280HQAG} 2010-01-02 06:59:00  PMB1928  PC-0026   
13       {X4J4-O2DE04LN-6069UNXT} 2010-01-02 07:09:00  DCB0446  PC-7490   
14       {L2L1-Y3CC03EE-4739DMKJ} 2010-01-02 07:13:59  JBI1134  PC-2226   
15       {L8S9-Q2NS95NB-3167LHVG} 2010-01-02 07:15:00  OHR0583  PC-0804   
16       {L0R0-D3NR92UX-9782DQLO} 2010-01-02 07:16:00  QGB0554  PC-2999   
17       {B3I3-N7WL96RV-0190VIKR} 2010-01-02 07:16:08  JBI1134  PC-2226   
18       {E9W5-L8HB85RJ-1921SJPX} 2010-01-02 07:17:00  HWR1573  PC-0766   
19       {Y8X6-P7RZ78IB-5529OKIL} 2010-01-02 07:19:00  RDC0128  PC-2973   
20       {O7D4-E1LN25FZ-3818EPVD} 2010-01-02 07:20:00  LAS0129  PC-1252   
21       {W2Y8-O5TX23UE-5273HELU} 2010-01-02 07:21:00  LIJ1581  PC-6639   
22       {H9L6-T5BK67RE-9499GXDI} 2010-01-02 07:22:00  BCC1597  PC-2558   
23       {Y9Z7-X7DP20UN-1901FNFC} 2010-01-02 07:24:00  AKS0586  PC-5335   
24       {P4C0-T4BM87QB-3449PLUA} 2010-01-02 07:24:00  HRT0984  PC-7416   
25       {U0U5-J2AX64NN-7430KKNP} 2010-01-02 07:27:00  VRP0267  PC-5687   
26       {I6W1-W3MG71AW-4776ITNU} 2010-01-02 07:28:00  AHN0681  PC-9843   
27       {W0Y4-M0TE51CP-5841GYLV} 2010-01-02 07:28:00  EMR0269  PC-7899   
28       {O2M9-M7AZ48CD-8197GBWM} 2010-01-02 07:29:00  DSB0530  PC-7866   
29       {T6R9-O7HI22LC-5988COVZ} 2010-01-02 07:30:00  BWJ1539  PC-4624   
...                           ...                 ...      ...      ...   
1810040  {G5I1-R0XJ07AC-7487XRNX} 2011-06-01 23:24:40  HJS0909  PC-3195   
1810041  {H7F4-U1EL57SI-9348VXDR} 2011-06-01 23:25:29  AAW0914  PC-2929   
1810042  {T5Z7-Z2MJ77KU-0801KOCI} 2011-06-01 23:26:48  AAW0914  PC-2929   
1810043  {W9P2-N7YB71XM-3787FVLI} 2011-06-01 23:26:56  RVJ0046  PC-3773   
1810044  {L0A8-S8OH07UD-8677WLEE} 2011-06-01 23:27:51  GCH0470  PC-4875   
1810045  {W3K4-P8SD30US-2480JFUC} 2011-06-01 23:28:45  HJS0909  PC-9775   
1810046  {G9N5-R1DS53QJ-5347LALO} 2011-06-01 23:30:02  JWS0047  PC-1624   
1810047  {H8M9-X4QA05YY-2616TSEI} 2011-06-01 23:30:25  GCH0470  PC-5439   
1810048  {F2A8-D2ZT53SD-2921JLSZ} 2011-06-01 23:34:36  LSH0464  PC-3195   
1810049  {G5K4-L4WZ39YO-9183MNAQ} 2011-06-01 23:38:10  GCH0470  PC-5439   
1810050  {N9H6-K5YG57LX-1513IOEI} 2011-06-01 23:39:22  JUW1135  PC-3180   
1810051  {O5Y9-R3CN80OK-0696LHSI} 2011-06-01 23:39:41  OQE0920  PC-2457   
1810052  {S5V8-A8VO38YN-7195WFEW} 2011-06-01 23:42:35  JWS0047  PC-1624   
1810053  {S5X8-G0ZU96EE-1140KTBE} 2011-06-01 23:43:13  JWS0047  PC-8360   
1810054  {K0P3-P9US34DN-0290GQMY} 2011-06-01 23:45:33  CAP0471  PC-8654   
1810055  {Z7U1-B9NW44DN-2720MNRB} 2011-06-01 23:46:27  JUW1135  PC-3180   
1810056  {V9V5-P7HP10IC-5992FMDY} 2011-06-01 23:46:37  HJS0909  PC-9775   
1810057  {S3Y7-O6UL72ST-2475WMWP} 2011-06-01 23:47:45  XAW0912  PC-1967   
1810058  {R5K5-O0BS23CH-6212QABH} 2011-06-01 23:49:18  XAW0912  PC-1967   
1810059  {O6U1-B3KX62DE-8699DEOZ} 2011-06-01 23:49:34  LSH0464  PC-3195   
1810060  {E7E0-S8RT34IT-1800RDHG} 2011-06-01 23:50:10  CAP0471  PC-8654   
1810061  {Q8Y6-Q9QC66JS-4091VMDE} 2011-06-01 23:56:46  J

In [47]:
def get_day(logon_df):
    logon_df['date'] = pd.to_datetime(logon_df.date, format='%m/%d/%Y %H:%M:%S')
    logon_df['day'] =logon_df['date'].dt.floor('D')

In [50]:
total_days = len(logon_df['day'])
print("total log of ", total_days, "days")
# len(logon_df[logon_df['day'] == "2011-06-01"])

total log of  1810070 days


In [49]:
days = logon_df['day'].unique()
file_df = pd.read_csv(dataset_dir/ 'file.csv')
http_df = pd.read_csv(dataset_dir / 'http.csv')
email_df = pd.read_csv(dataset_dir / 'email.csv')
device_df = pd.read_csv(dataset_dir / 'device.csv')

get_day(file_df)
get_day(http_df)
get_day(email_df)
get_day(device_df)

# Summary of all
|  total user | senerio 1 | senerio 2 | senerio 3 | senerio 4| Normal user |
| --- | --- | --- | -- | -- | -- |
| 2000 | 29 | 30 | 10 | 30 | 1901 |

| logs per * | senerio 1 | senerio 2 | senerio 3 | senerio 4| Normal user | total |
| --- | --- | --- | -- | -- | -- | -- |
|total logs | 486 | 6477 | 184 | 3181| 79846376 | 79856704

| logs per * | senerio 1 | senerio 2 | senerio 3 | senerio 4| Normal user | total |
| --- | --- | --- | -- | -- | -- | -- |
|logs per user | 17 | 215 | 18 | 106| 42002 | 39928